[Feature] CAS Institution Login [#CAS-3]

cslzchen opened the pull request:
## Tickets and related PRs

Ticket: https://openscience.atlassian.net/browse/CAS-3
OSF-PR: https://github.com/CenterForOpenScience/osf.io/pull/6082
## Description

Use CAS for institution login.
## Deployment Considerations

```
cas.osf.login.url=/login?
cas.institution.login.url=/login?campaign=institution
```
## Changes
### UI
- Slightly-modified CAS Login Page

![cas-login-page](https://cloud.githubusercontent.com/assets/3750414/19245060/3b44de10-8eed-11e6-8fe6-50245ac4009e.png)
- New CAS Institution Login Page

![cas-institution-login-page](https://cloud.githubusercontent.com/assets/3750414/19245073/4421f1a8-8eed-11e6-867d-c47309e8a3ba.png)
- Old OSF Institution Login Page

![osf-inst-login-old](https://cloud.githubusercontent.com/assets/3750414/17342565/e395af10-58c7-11e6-8218-a5328b4625d2.png)
### CAS Backend
- Refactor `OpenScienceFrameworkLogoutHandler` to `OpenScienceFrameworkInstitutionHandler`:
  add necessary institution fields;
  add `getInstitutionLoginUrls` and return a map of institutions' name and login url.
- **Action State**: `beforeLogin`
  **Action Event**: `OpenScienceFrameworkLoginHandler.beforeLogin`
  A decision state that chooses between normal OSF login and institution login, which is also scalable for campaign login.
- **Action State**: `institutionLogin`
  **Action Event**: `OpenScienceFrameworkInstitutionLoginHandler.getInstitutions`
  Retrieve institutions from DB and put a map in flow.
-  **View State**: `viewInstitutionLoginForms`
  Target alternative login page `casInstitutionLoginView.jsp`
- Modified CAS login flow:

```
    -->> remoteAuthenticate
        -->> success: sendTGT
        -->> failure: authFailure
        -->> error: beforeLogin

   -->> beforeLogin
        -->> osf: viewLoginForm
        -->> institution: institutionLogin

    -->> institutionLogin
        -->> success: viewInstitutionLoginForms
```
### OSF Backend

Please refer to [OSF-PR#6082](https://github.com/CenterForOpenScience/osf.io/pull/6082) for more information.
## Side Effects

No

cslzchen committed:
Add institution login page for CAS + understand CAS login flow.

cslzchen committed:
Refactor logoutHandler to institutionAuthenticationHandler.

cslzchen committed:
Refactor OpenScienceFrameworkInstitution to add all institution fields.

cslzchen committed:
Implement Institution Login from CAS.
    - new action state: institutionLogin
        - class: OpenScienceFrameworkLoginThroughInstitutions
        - action: get instituions (name and login url) and add the map to flow for view state below
    - new view state: viewInstitutionLoginForm
    - take care of several redirection issues

cslzchen committed:
Implement UI for CAS Institution Login (Part 1)
    - mimic functionality of OSF Institution login page
    - update cas.css for improved UI exprience
    - handle no service redirection
TODO (Part 2):
    - foward service URL to institution login
    - fix links on page
    - fix redirection

cslzchen committed:
Replace `institution=true` with `campaign=institution` and several fixes:
    - links for terms, policy and cookie
    - comments and java docs

cslzchen committed:
Implement UI for CAS Institution Login (Part 2)
    - add new institution and non-institution login link
    - foward service URL to institution login for correct redirection
    - improve css

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/adaptors/mongodb/OpenScienceFrameworkLogoutHandler.java:
```diff
@@ -1,118 +0,0 @@
-/*
```
TO CR: this file is renamed to `OpenScienceFrameworkInstitutionAuthenticationHandler`. The one commit for the rename consider it `mv` but the PR does not. There are a two main changes: 1. add all institution related fields to `OpenScienceFrameworkInstitution` 2. add a method to return a map of institution name and login url.

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginThroughInstitutions.java:
```diff
@@ -0,0 +1,111 @@
+/*
```
To CR: for the new `action-state: institutionLogin`, triggered by `Event(institution)` in `action-state: remoteAuthenticate` and go on to the new `view-state: viewInstitutionLoginForm`.

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkTerminateSessionAction.java:
```diff
@@ -130,9 +130,9 @@ public Event terminate(final RequestContext context) {
+            final String institutionLogoutUrl = this.institutionHandler.findInstitutionLogoutUrlById(institutionId);
             if (institutionLogoutUrl == null) {
-                logger.warn("Institution {} does not have logout url, use default logout redirection instead", institutionId);
+                logger.warn("Institution {} does not have an dedicated logout url, use default logout redirection instead", institutionId);
```
To CR: I will fix this typo, should be `a dedicated`.

cslzchen commented on etc/cas.properties:
```diff
@@ -32,10 +32,15 @@ google.analytics.id=
+#
+cas.osf.login.url=/login?
+cas.institution.login.url=/login?campaign=institution
+
```
For CR: need to update this in deployment.

icereval replied:
if not already done, please note in the pr description

icereval replied:
e.g. https://github.com/CenterForOpenScience/modular-file-renderer/pull/223

cslzchen committed:
Fix conflicts and merge remote-tracking branch:
'upstream/develop' >> feature/cas-institution-login.

cslzchen committed:
Add button hover and click effect for ORCID login button.

cslzchen committed:
Improve institution login page.

cslzchen committed:
Improve top and bottom jsp.

cslzchen committed:
Add hover/active effect for sign in button.

cslzchen committed:
Fix typo and remove comments.

icereval commented on cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:
```diff
@@ -167,12 +167,23 @@ public String getInstitutionId() {
+     */
+    protected boolean isInstitution(final RequestContext context) {
+        final String campaign = context.getRequestParameters().get("campaign");
+        return campaign != null && "institution".equals(campaign);
```
may not need `campaign != null &&`

icereval commented on cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginThroughInstitutions.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    /**
+     * Sort a Map by value. Return teh sorted Map.
```
typo `teh`

icereval commented on cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginThroughInstitutions.java:
```diff
@@ -0,0 +1,108 @@
+        final Map<String, String> sortedInstitutions = sortByValue(institutions);
+        logger.info(String.format("Institutions loaded: %s", sortedInstitutions.toString()));
+        context.getFlowScope().put("institutions", sortedInstitutions);
+        String target = "";
```
should target default to encoded `/login`?

icereval commented on cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkLoginThroughInstitutions.java:
```diff
@@ -0,0 +1,108 @@
+     * @param <V> The Type of the Value
+     * @return Map&lt;K,V&gt;
+     */
+    private static <K, V extends Comparable<? super V>> Map<K, V> sortByValue(final Map<K, V> map) {
```
is this utility method from somewhere we can source?

cslzchen replied:
No reliable package found.

icereval commented on cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkTerminateSessionAction.java:
```diff
@@ -63,32 +63,32 @@
     private final CookieRetrievingCookieGenerator warnCookieGenerator;
 
-    /** Logout handler. */
+    /** Institution authentication handler for logout. */
```
Institution handler

cslzchen committed:
Remove unused property and getters/setters.

icereval commented on cas-server-support-osf/src/main/java/io/cos/cas/adaptors/mongodb/OpenScienceFrameworkInstitutionAuthenticationHandler.java:
```diff
@@ -0,0 +1,221 @@
+ * @author Longze Chen
+ * @since 4.1.0
+ */
+public class OpenScienceFrameworkInstitutionAuthenticationHandler {
```
OpenScienceFrameworkInstitutionHandler

icereval commented on cas-server-webapp/src/main/webapp/WEB-INF/spring-configuration/institutionAuthenticationManager.xml:
```diff
@@ -22,8 +22,12 @@
        xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-3.2.xsd">
 
-    <bean id="openScienceFrameworkLogoutHandler" class="io.cos.cas.adaptors.mongodb.OpenScienceFrameworkLogoutHandler"
+    <bean id="openScienceFrameworkInstitutionAuthenticationHandler" class="io.cos.cas.adaptors.mongodb.OpenScienceFrameworkInstitutionAuthenticationHandler"
```
openScienceFrameworkInstitutionHandler

icereval commented on cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/casInstitutionLoginView.jsp:
```diff
@@ -0,0 +1,65 @@
+    <br>
+    <script>
+        function institutionLogin () {
+            var form = document.getElementById("institution-form-select");
```
trivial, use single quotes in JS when possible.

https://github.com/airbnb/javascript/blob/master/README.md#strings--quotes

icereval commented on cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/casInstitutionLoginView.jsp:
```diff
@@ -0,0 +1,65 @@
+    </section>
+    <br>
+    <script>
+        function institutionLogin () {
```
move JS to backend `getInstitutions` method

cslzchen replied:
For a `(KEY, VALUE)` pair in `<form:select items="${MAP}" />`, the following dropdown is generated: `<option value="KEY">VALUE</option>`, where `KEY` is the hidden login url and `VALUE` is the display name. Need to build the `MAP` with `target` in `KEY`.

icereval commented on cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/bottom.jsp:
```diff
@@ -21,16 +21,25 @@
+        <c:choose>
+            <c:when test="${campaign == 'INSTITUTION'}">
+                <spring:eval var="osfLoginUrl" expression="@casProperties.getProperty('cas.osf.login.url')" />
+                <a id="alternative-osf" href="${osfLoginUrl}${not empty param.service ? 'service=' : ''}${fn:escapeXml(param.service)}">Back to Non-institution Login</a>&nbsp;&nbsp;
```
<strike>Back to </strike>

icereval commented on cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/bottom.jsp:
```diff
@@ -21,16 +21,25 @@
+            </c:when>
+            <c:otherwise>
+                <spring:eval var="institutionLoginUrl" expression="@casProperties.getProperty('cas.institution.login.url')" />
+                <a id="aleternative-institution" href="${institutionLoginUrl}${not empty param.service ? '&service=' : ''}${fn:escapeXml(param.service)}">Login through Your Institution</a>&nbsp;&nbsp;
```
change to flow, no longer going over to OSF page

cslzchen replied:
Already changed to CAS institution login. I will leave as it is now.

icereval commented on cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/top.jsp:
```diff
@@ -28,32 +28,47 @@
+                    <c:if test="${campaign eq 'INSTITUTION'}">
+                        <div id="description">
+                            <br><br><spring:message code="screen.institution.login.message.a" />
+                            <spring:message code="screen.institution.login.message.b" />
```
`.a/b`?

icereval commented on cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/top.jsp:
```diff
@@ -28,32 +28,47 @@
+                    <span id="title">
+                        <span class="title-full">Open&nbsp;Science&nbsp;Framework</span>
+                        <span class="title-abbr">Open&nbsp;Science<br>Framework</span>
+                        <c:if test="${campaign eq 'INSTITUTION'}"><span> for&nbsp;Institutions</span></c:if>
```
use `OSF Institutions`

icereval replied:
should be able to remove resize exception

icereval commented on cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/top.jsp:
```diff
@@ -28,32 +28,47 @@
+                    </span>
+                </div>
+                <div class="responsive">
+                    <c:if test="${campaign eq 'INSTITUTION'}">
```
In general campaign should be a data structure passed to the front end, eventually surfaced from the database.

cslzchen replied:
Agreed. Leave it as it is now. Add this change for https://github.com/CenterForOpenScience/cas-overlay/pull/24.

icereval commented on cas-server-webapp/src/main/webapp/css/cas.css:
```diff
@@ -249,18 +282,15 @@ footer a:link, footer a:visited {
-#fm1 .row .btn-oauth,
-#fm1 .row .btn-oauth:hover,
-#fm1 .row .btn-oauth:active {
+#fm1 .row .btn-oauth {
```
test in safari,ie,firefox,chrome

icereval reviewed (commented):
first pass

icereval commented on cas-server-webapp/src/main/webapp/WEB-INF/webflow/login/login-webflow.xml:
```diff
@@ -97,8 +97,17 @@
         <transition on="success" to="sendTicketGrantingTicket" />
         <transition on="authenticationFailure" to="handleAuthenticationFailure" />
         <transition on="error" to="viewLoginForm" />
+        <transition on="institution" to="institutionLogin" />
```
adjust `error` state to send to a campaign decision step.

cslzchen replied:
Add `OpenScienceFrameworkLoginHandler` with action `beforeLogin` which transits to `osfLogin` or `institutionLogin`. In the future, there will be another `osfCampaignLogin` to handle campaigns except institution in general.

cslzchen committed:
Several fixes from CR:
    - OpenScienceFrameworkInstitutionHandler (Rename)
    - Remove redundant null check for string equal
    - Set default target to '/login'
    - Fix Typo

cslzchen committed:
Add login decision action state `beforeLogin`:
    - remoteAuthenticate transits to beforeLogin on error
    - beforeLogin decide normal osf login or institution login

cslzchen committed:
Handle target redirection in flow instead of page.
    - OpenScienceFrameworkInstituionLoginHandler (rename) builds target into login url
    - getInstitutionLoginUrls (rename) takes target when building map

cslzchen committed:
Improve page title and message, fix style check.

cslzchen commented:
@icereval ready to go :)

cslzchen committed:
Add back cookie usage message.
